karmada-search: Fix lock race affects watch RestChan not close, causing client watch api to hang.

xigang opened the pull request:
**What type of PR is this?**

/kind bug

<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:
When the `result` channel of `watchMux` is blocked, it will cause `goroutine` leak and lock race. this problem affects the client watch api interface to hang when consuming `watch.RestChan`.


fetch goroutine stack:
```
curl http://localhost:6062/debug/pprof/goroutine?debug=1
```

Through the `goroutine` stack below, there are `1892` goroutines blocked in `runtime_SemacquireMutex`.

```
1892 @ 0x439936 0x44a80c 0x44a7e6 0x466ba5 0x1a1a5fd 0x1a1a5de 0x1a1a32a 0x46acc1
#       0x466ba4        sync.runtime_SemacquireMutex+0x24                                                                               /usr/lib/golang/src/runtime/sema.go:71
#       0x1a1a5fc       sync.(*RWMutex).RLock+0x7c                                                                                      /usr/lib/golang/src/sync/rwmutex.go:63
#       0x1a1a5dd       go/src/github.com/karmada/pkg/search/proxy/store.(*watchMux).startWatchSource.func1+0x5d     /go/src/go/src/github.com/karmada/pkg/search/proxy/store/util.go:229
#       0x1a1a329       go/src/github.com/karmada/pkg/search/proxy/store.(*watchMux).startWatchSource+0xe9           /go/src/go/src/github.com/karmada/pkg/search/proxy/store/util.go:237

967 @ 0x439936 0x405565 0x40511d 0x1a1a691 0x1a1a32a 0x46acc1
#       0x1a1a690       go/src/github.com/karmada/pkg/search/proxy/store.(*watchMux).startWatchSource.func1+0x110    /go/src/go/src/github.com/karmada/pkg/search/proxy/store/util.go:235
#       0x1a1a329       go/src/github.com/karmada/pkg/search/proxy/store.(*watchMux).startWatchSource+0xe9           /go/src/go/src/github.com/karmada/pkg/search/proxy/store/util.go:237

573 @ 0x439936 0x44a80c 0x44a7e6 0x466ba5 0x474fa5 0x476576 0x476555 0x1a1a132 0x18e1a0d 0x18ce325 0x16aadc2 0x18cd8fb 0x79d3cf 0x16acb6c 0x79d3cf 0x1a3cbcb 0x79d3cf 0x18da0a6 0x16aadc2 0x18d9a1b 0x18f6def 0x16ac4a2 0x16976f1 0x19214e6 0x19214b6 0x16be6a5 0x79d3cf 0x16c807d 0x79d3cf 0x16be098 0x79d3cf 0x186a2e4
#       0x466ba4        sync.runtime_SemacquireMutex+0x24                                                                               /usr/lib/golang/src/runtime/sema.go:71
#       0x474fa4        sync.(*Mutex).lockSlow+0x164                                                                                    /usr/lib/golang/src/sync/mutex.go:138
#       0x476575        sync.(*Mutex).Lock+0x35                                                                                         /usr/lib/golang/src/sync/mutex.go:81
#       0x476554        sync.(*RWMutex).Lock+0x14                                                                                       /usr/lib/golang/src/sync/rwmutex.go:111
#       0x1a1a131       go/src/github.com/karmada/pkg/search/proxy/store.(*watchMux).Stop+0x51                       /go/src/go/src/github.com/karmada/pkg/search/proxy/store/util.go:193
#       0x18e1a0c       k8s.io/apiserver/pkg/endpoints/handlers.serveWatch+0xf8c                                                        /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:139
#       0x18ce324       k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1.2+0x64                                               /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go:270
#       0x16aadc1       k8s.io/apiserver/pkg/endpoints/metrics.RecordLongRunning+0x721                                                  /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:427
#       0x18cd8fa       k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1+0xf1a                                                /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go:269
#       0x79d3ce        net/http.HandlerFunc.ServeHTTP+0x2e                                                                             /usr/lib/golang/src/net/http/server.go:2046
#       0x16acb6b       k8s.io/apiserver/pkg/endpoints/metrics.InstrumentHandlerFunc.func1+0x2cb                                        /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:509
#       0x79d3ce        net/http.HandlerFunc.ServeHTTP+0x2e                                                                             /usr/lib/golang/src/net/http/server.go:2046
#       0x1a3cbca       go/src/github.com/karmada/pkg/controllers/proxy.(*Controller).Connect.func1+0x6aa        /go/src/go/src/github.com/karmada/pkg/controllers/proxy/proxy_controller.go:256
#       0x79d3ce        net/http.HandlerFunc.ServeHTTP+0x2e                                                                             /usr/lib/golang/src/net/http/server.go:2046
#       0x18da0a5       k8s.io/apiserver/pkg/endpoints/handlers.ConnectResource.func1.1+0x245                                           /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:224
#       0x16aadc1       k8s.io/apiserver/pkg/endpoints/metrics.RecordLongRunning+0x721                                                  /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:427
#       0x18d9a1a       k8s.io/apiserver/pkg/endpoints/handlers.ConnectResource.func1+0xcba                                             /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:218
#       0x18f6dee       k8s.io/apiserver/pkg/endpoints.restfulConnectResource.func1+0x6e                                                /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/installer.go:1228
#       0x16ac4a1       k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1+0x2e1                                          /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:484
#       0x16976f0       github.com/emicklei/go-restful.(*Container).dispatch+0x9d0                                                      /go/src/go/src/github.com/karmada/vendor/github.com/emicklei/go-restful/container.go:288
#       0x19214e5       github.com/emicklei/go-restful.(*Container).Dispatch+0x645                                                      /go/src/go/src/github.com/karmada/vendor/github.com/emicklei/go-restful/container.go:199
#       0x19214b5       k8s.io/apiserver/pkg/server.director.ServeHTTP+0x615                                                            /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/server/handler.go:146
#       0x16be6a4       k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1+0x1a4                                         /go/src/go/src/github.com/karmada/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:103
```

Refer to [cacheWatcher](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L53C9-L53C9) send event to input channel to implement adding a timeout.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L216

```
	// OK, block sending, but only until timer fires.
	select {
	case c.input <- event:
		return true
	case <-timer.C:
		closeFunc()
		return false
	}
```


**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
@ikaven1024 @RainbowMango @XiShanYongYe-Chang 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
`karmada-search`: Fix lock race affects watch RestChan not close, causing client watch api to hang.
```

xigang commented:
/cc @ikaven1024 @RainbowMango @XiShanYongYe-Chang

XiShanYongYe-Chang commented:
Thanks @xigang 
/assign

RainbowMango commented:
@xigang Do you know how to reproduce it? 

> When the result channel of watchMux is blocked, it will cause goroutine leak and lock race.

Since the [result channel](https://github.com/karmada-io/karmada/blob/d6e28817c381b7524ac0cfabd077dd25226f9410/pkg/search/proxy/store/util.go#L177) is no cached, so I can understand that the goroutine will be blocked when trying send event to it:
https://github.com/karmada-io/karmada/blob/d6e28817c381b7524ac0cfabd077dd25226f9410/pkg/search/proxy/store/util.go#L259

Can you remind me why it will cause goroutine leak and lock race?

xigang commented:
Currently, our federated cluster has 1.8W+ nodes (the size of the nodes will increase). The client uses the `NodeInformer` ListWatch Nodes resource for a period of time (more than 10 hours), and the client's watch interface will hang.

xigang commented:
> @xigang Do you know how to reproduce it?
> 
> > When the result channel of watchMux is blocked, it will cause goroutine leak and lock race.
> 
> Since the [result channel](https://github.com/karmada-io/karmada/blob/d6e28817c381b7524ac0cfabd077dd25226f9410/pkg/search/proxy/store/util.go#L177) is no cached, so I can understand that the goroutine will be blocked when trying send event to it:
> 
> https://github.com/karmada-io/karmada/blob/d6e28817c381b7524ac0cfabd077dd25226f9410/pkg/search/proxy/store/util.go#L259
> 
> Can you remind me why it will cause goroutine leak and lock race?

[client side]
First, we added logs to the [watchHandler](https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L729) of the `reflector` of client-go and found that the client was stuck `w.ResultChan()`.

```
	for {
		select {
		case <-stopCh:
			klog.Infof("reflector stopped")
			return errorStopRequested
		case err := <-errc:
			klog.Infof("watch handler received error: %v", err.Error())
			return err
		case event, ok := <-w.ResultChan():
			if !ok {
				klog.Infof("watch server result chan is closed. for loop")
				break loop
			}
			klog.Infof("consume event success")
```

reproduce log:
 
![image](https://github.com/karmada-io/karmada/assets/3377677/a36a720d-8462-4f0c-a84f-06970e374354)

[server side]
Then we added debug log for [serveWatch](https://github.com/kubernetes/kubernetes/blob/5ce0bd95cc6e6cf0333f6e58623344d9562ba7e2/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L66) to` WatchServer` on the server side, and found that when the watch http2 connection exited, it got stuck while executing `defer watcher.Stop()`.

```
// serveWatch will serve a watch response.
// TODO: the functionality in this method and in WatchServer.Serve is not cleanly decoupled.
func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions negotiation.MediaTypeOptions, req *http.Request, w http.ResponseWriter, timeout time.Duration) {
	klog.Infof("watcher %v connecting", req.UserAgent())
	defer func(){
		klog.Infof("watcher %v stop done in serveWatch", req.UserAgent()) //No execution here
	}()
	defer watcher.Stop()
	defer func(){
		klog.Infof("watcher %v http2 closed.", req.UserAgent()) //The code is executed here
	}()

```

Finally, through the goroutine profile, it was found that a large number of goroutines were stuck and locked `runtime_SemacquireMutex`.

xigang commented:
> > ![image](https://user-images.githubusercontent.com/3377677/282004834-a36a720d-8462-4f0c-a84f-06970e374354.png)
> 
> Why dose it block at fourth renow, not before?

@ikaven1024 In the log, the hang appears not for the fourth watch, but for the Nth time. Because the server will regularly close the watch api connection, the client will renew the watch. This prevents the client from always watching a single `kube-apiserver`.

xigang commented:
@ikaven1024 Thanks for the analysis. Keep `RLock`. Is it safer to add a timeout?

xigang commented:
> > Is there any problem in remove `RLock` when multiple goroutines process `w.done` and `w.result`?
> 
> It's safe reading, writing and closing a chan in multiple goroutines.
> 
> > If RLock is retained. Is it better to use timeout to reduce lock race?
> 
> It will cause client break the watch, and send a new watch request, Increasing the burden on the server.

@ikaven1024  okay, I have resubmitted the code, please review.

xigang commented on pkg/search/proxy/store/util.go:
```diff
@@ -250,13 +251,10 @@ func (w *watchMux) startWatchSource(source watch.Interface, decorator func(watch
 				return
-			default:
-				w.result <- copyEvent
+			case w.result <- copyEvent:
```
@ikaven1024 code updated. Ask for help, the unit test did not pass, and I couldn't find what went wrong.😄

xigang commented on pkg/search/proxy/store/util.go:
```diff
@@ -225,6 +225,7 @@ func (w *watchMux) Stop() {
 }
 
 func (w *watchMux) startWatchSource(source watch.Interface, decorator func(watch.Event)) {
+	defer func() { _ = recover() }()
```
Yes, This avoids writing panic after w.result is closed.

RainbowMango replied:
This recover() suppresses panic forever, can you explain why we need to do that?

> This avoids writing panic after w.result is closed.

Can we eliminate this risk?

My concern about the `recover()` is it just suppresses issues, might bring debug effort.

xigang replied:
> > Yes, This avoids writing panic after w.result is closed.
> 
> I mean we should print log when panic.

panic log has been added.

xigang replied:
@RainbowMango The reason for adding `recover()` is that I am worried that multiple gourinue write the `w.result` channel at the same time. When one goroutine close `w.result`, other goroutines continue to write` w.result`, triggering panic.

RainbowMango replied:
Write to a closed channel won't panic in `select case`, like https://github.com/karmada-io/karmada/pull/4212/files#diff-688048f0ec4cf2d213f3bd0437bde2531b124c71d172319b20f25cd140fc238cR255.

XiShanYongYe-Chang commented on pkg/search/proxy/store/util.go:
```diff
@@ -166,22 +167,47 @@ func (c *multiClusterContinue) String() string {
-		done:   make(chan struct{}),
+		result:         make(chan watch.Event, 100),
+		done:           make(chan struct{}),
+		timeoutFactory: &realTimeoutFactory{timeout: time.Second * 10},
```
After we remove the read lock, do we still need to add a timeout?

xigang replied:
@ikaven1024 @XiShanYongYe-Chang I updated a version of the code. If `RLock` is removed, data race for `w.result` will be triggered.

unit test log:
```
=== RUN   Test_watchMux_StopBySelf
==================
WARNING: DATA RACE
Write at 0x00c00093af70 by goroutine 331:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  github.com/karmada-io/karmada/pkg/search/proxy/store.(*watchMux).Stop()
      /Users/wangxigang/go/src/github.com/xigang/karmada/pkg/search/proxy/store/util.go:224 +0x10c
  github.com/karmada-io/karmada/pkg/search/proxy/store.(*watchMux).Stop-fm()
      <autogenerated>:1 +0x33

Previous read at 0x00c00093af70 by goroutine 328:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:160 +0x0
  github.com/karmada-io/karmada/pkg/search/proxy/store.(*watchMux).startWatchSource()
      /Users/wangxigang/go/src/github.com/xigang/karmada/pkg/search/proxy/store/util.go:255 +0x3c8
  github.com/karmada-io/karmada/pkg/search/proxy/store.(*watchMux).Start.func1()
      /Users/wangxigang/go/src/github.com/xigang/karmada/pkg/search/proxy/store/util.go:200 +0x5d

Goroutine 331 (running) created at:
  time.goFunc()
      /usr/local/go/src/time/sleep.go:176 +0x44

Goroutine 328 (running) created at:
  github.com/karmada-io/karmada/pkg/search/proxy/store.(*watchMux).Start()
      /Users/wangxigang/go/src/github.com/xigang/karmada/pkg/search/proxy/store/util.go:200 +0x56
  github.com/karmada-io/karmada/pkg/search/proxy/store.Test_watchMux_StopBySelf()
      /Users/wangxigang/go/src/github.com/xigang/karmada/pkg/search/proxy/store/util_test.go:38 +0x34e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1648 +0x44
==================
    testing.go:1465: race detected during execution of test
--- FAIL: Test_watchMux_StopBySelf (1.00s)
```

xigang replied:
> This `timeout` is incompatibility with the one in `serveWatch`:
> 
> https://github.com/karmada-io/karmada/blob/3f5c9073ab197dbc83f217448215a767f7fd0e94/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L253-L259

Yes, This two-part timeout is incompatible.

XiShanYongYe-Chang commented:
Hi @xigang Is the current pr based on fixing #4221?

xigang commented:
> Hi @xigang Is the current pr based on fixing #4221?

@XiShanYongYe-Chang Yes, @ikaven1024 open a new PR [#4221](https://github.com/karmada-io/karmada/pull/4221) to fix this bug.

XiShanYongYe-Chang commented:
Thanks for your quick response @xigang

RainbowMango commented:
> @XiShanYongYe-Chang Yes, @ikaven1024 open a new PR https://github.com/karmada-io/karmada/pull/4221 to fix this bug.

We usually don't open another PR to replace the current in-review PR. I don't know which PR we should move forward with.

@xigang What do you think?

xigang commented:
> > @XiShanYongYe-Chang Yes, @ikaven1024 open a new PR #4221 to fix this bug.
> 
> We usually don't open another PR to replace the current in-review PR. I don't know which PR we should move forward with.
> 
> @xigang What do you think?

@RainbowMango Agreed, in order to track this issue, let's continue to fix this issue in this PR.

RainbowMango commented:
Yes, we can focus on fixing this issue and keep the minimum change( so that we can easily backport this to release branches), and let #4221 focus on additional improvement.

Does that make sense to you? @xigang @ikaven1024

xigang commented:
> Yes, we can focus on fixing this issue and keep the minimum change( so that we can easily backport this to release branches), and let #4221 focus on additional improvement.
> 
> Does that make sense to you? @xigang @ikaven1024

@ikaven1024 Do I migrate [@4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR?

xigang commented:
> Yes, we can focus on fixing this issue and keep the minimum change( so that we can easily backport this to release branches), and let #4221 focus on additional improvement.
> 
> Does that make sense to you? @xigang @ikaven1024

@ikaven1024 [#4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR? Or use timeout to temporarily fix this issue?

xigang commented:
> > @ikaven1024 Migrate [#4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR? Or use timeout to temporarily fix this issue?
> 
> You can migrate #4221 here, and i will close it.



> > @ikaven1024 Migrate [#4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR? Or use timeout to temporarily fix this issue?
> 
> You can migrate #4221 here, and i will close it.



> > @ikaven1024 Migrate [#4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR? Or use timeout to temporarily fix this issue?
> 
> You can migrate #4221 here, and i will close it.

Done.

xigang commented:
> > @ikaven1024 Migrate [#4221](https://github.com/karmada-io/karmada/pull/4221) code to this PR? Or use timeout to temporarily fix this issue?
> 
> You can migrate #4221 here, and i will close it.

Done.

RainbowMango commented on pkg/search/proxy/store/util.go:
```diff
@@ -195,9 +195,21 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
+	wg := sync.WaitGroup{}
 	for _, source := range w.sources {
-		go w.startWatchSource(source.watcher, source.decorator)
+		source := source
```
> Reports declarations of variables that overlap with the declarations in the outer scope.

A warning from Goland IDE.

If you want to `pin` the variable, we can give it another name, or:
```diff
diff --git a/pkg/search/proxy/store/util.go b/pkg/search/proxy/store/util.go
index 00f93c467..dbb765959 100644
--- a/pkg/search/proxy/store/util.go
+++ b/pkg/search/proxy/store/util.go
@@ -196,8 +196,8 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
 // Start run the watcher
 func (w *watchMux) Start() {
        wg := sync.WaitGroup{}
-       for _, source := range w.sources {
-               source := source
+       for i := range w.sources {
+               source := w.sources[i]
                wg.Add(1)
                go func() {
                        defer wg.Done()


```

xigang replied:
@RainbowMango Fixed.

RainbowMango reviewed (commented):
/lgtm

/assign @ikaven1024 
For the final decision.

XiShanYongYe-Chang commented:
Hi @xigang do you think we need to cherry-pick this patch to the previous branch?

xigang commented:
> cherry-pick

@XiShanYongYe-Chang I think we should cherry-pick the previous branch and let me handle it.

XiShanYongYe-Chang commented:
> @XiShanYongYe-Chang I think we should cherry-pick the previous branch and let me handle it.

Thanks~
